sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+
+/**
+ * This aims to handle a nested column aliasing pattern inside the `ColumnPruning` optimizer rule.
+ * If a project or its child references to nested fields, and not all the fields
+ * in a nested attribute are used, we can substitute them by alias attributes; then a project
+ * of the nested fields as aliases on the children of the child will be created.
+ */
+object NestedColumnAliasing {
+
+  def unapply(plan: LogicalPlan)
+    : Option[(Map[GetStructField, Alias], Map[ExprId, Seq[Alias]])] = plan match {
+    case Project(projectList, child)
+        if SQLConf.get.nestedSchemaPruningEnabled && canProjectPushThrough(child) =>
```

Comment:
Hey @dongjoon-hyun, @viirya, and @dbtsai. Sorry for late input - I was just looking through the PRs in my notification queue.

Just out of curiosity. WDYT about this idea?

1. Putting this as a separate rule under `defaultBatches` in `Optimizer` 
2. Removing `spark.sql[.optimizer].serializer.nestedSchemaPruning.enabled`
3. Add the rule into `spark.sql.optimizer.excludedRules` as default.
4. Add some comments on the rule that this rule is intentionally excluded from optimizer as it's an experimental rule set.

In this way, we don't have to add a configuration for each nested column access improvement, and leverage the existing configuration. It requires to take a look here and there but I was thinking it's a cleaner way.

Reply:
Hi, @HyukjinKwon . Actually, this is the 3rd try for this. At every PR, there exists the same questions about `a separate rule`. :)

1. 1st PR: https://github.com/apache/spark/pull/23542#discussion_r248095908
2. 2nd PR: https://github.com/apache/spark/pull/23873#discussion_r261768034
3. 3rd one: Your.

In short, this should be here. Otherwise, the projection is already moved down as the whole attribute instead of a subset of fields.

Reply:
Oops, sorry looks I missed the discussion. but still we could remove other rule-specific configuration (not this PR one).. but yes, maybe I should go to that PR and ask the idea :).